UefiCpuPkg/Library/MpInitLib/DxeMpLib.c

```diff
@@ -565,17 +512,59 @@ RemoveNxprotection (
   EFI_STATUS                       Status;
   EFI_GCD_MEMORY_SPACE_DESCRIPTOR  MemDesc;
 
-  //
-  // TODO: Check EFI_MEMORY_XP bit set or not once it's available in DXE GCD
-  //       service.
-  //
   Status = gDS->GetMemorySpaceDescriptor (BaseAddress, &MemDesc);
   if (!EFI_ERROR (Status)) {
-    gDS->SetMemorySpaceAttributes (
-           BaseAddress,
-           Length,
-           MemDesc.Attributes & (~EFI_MEMORY_XP)
-           );
+    if ((MemDesc.Capabilities & EFI_MEMORY_XP) == EFI_MEMORY_XP ) {
+      if ((MemDesc.Attributes & EFI_MEMORY_XP) == EFI_MEMORY_XP) {
+        Status = gDS->SetMemorySpaceAttributes (
+                        BaseAddress,
+                        Length,
+                        MemDesc.Attributes & (~EFI_MEMORY_XP)
+                        );
+      }
+    }
+  }

```

Comment:
Is there a reason not to combine these? Both blocks lead to a single code path.

```suggestion
    if (((MemDesc.Capabilities & EFI_MEMORY_XP) == EFI_MEMORY_XP) && ((MemDesc.Attributes & EFI_MEMORY_XP) == EFI_MEMORY_XP)) {
      Status = gDS->SetMemorySpaceAttributes (
                      BaseAddress,
                      Length,
                      MemDesc.Attributes & (~EFI_MEMORY_XP)
                      );
    }
  }
```

If you find the existing way more readable, feel free to leave it. I find it a little more readable to have less nesting.

Reply:
Implemented in the latest.